src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncEHTests.cs

```diff
@@ -2104,5 +2104,148 @@ static void Main()
 ";
             CompileAndVerify(source, expected);
         }
+
+        [Fact, WorkItem(67091, "https://github.com/dotnet/roslyn/issues/67091")]
+        public void NestedCatch_DuplicateLocal()
+        {
+            var source = """
+                using System;
+                using System.Threading.Tasks;
+                using static System.Console;
+
+                class C
+                {
+                    static async Task Main()
+                    {
+                        await new C().M1();
+                    }
+
+                    int _counter;
+
+                    bool F(Exception ex)
+                    {
+                        _counter++;
+                        WriteLine($"F: {_counter} {ex.Message}");
+                        return _counter == 2;
+                    }
+
+                    async Task M1()
+                    {
+                        try
+                        {
+                            throw new Exception("M1");
+                        }
+                        catch (Exception ex) when (F(ex))
+                        {
+                            await M2(ex);
+                        }
+                        catch (Exception ex) when (F(ex))
+                        {
+                            try
+                            {
+                                throw new Exception("M1-catch2");
+                            }
+                            catch 
+                            {
+                                await M2(ex);
+                            }
+                        }
+                    }
+
+                    async Task M2(Exception ex)
+                    {
+                        WriteLine($"M2: {ex.Message}");
+                        await Task.Yield();
+                    }
+                }
+                """;
+            var expectedOutput = """
+                F: 1 M1
+                F: 2 M1
+                M2: M1
+                """;
+
+            CompileAndVerify(source, options: TestOptions.DebugExe, expectedOutput: expectedOutput).VerifyDiagnostics();
+            CompileAndVerify(source, options: TestOptions.ReleaseExe, expectedOutput: expectedOutput).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(67091, "https://github.com/dotnet/roslyn/issues/67091")]
+        public void NestedCatch_DuplicateLocal_Level2()
```

Comment:
>NestedCatch_DuplicateLocal_Level2 [](http://example.com/codeflow?start=20&length=33)

Is this test going to fail if we do not go through the entire parent chain in ```TryGetHoistedLocal```, but instead only check in an immediate parent?

Reply:
Yes, I have tried that.